Support a date time in intermediate timer catch events non-/interrupting timer boundary events

Wolgo opened the pull request:
## Description
For in intermediate, non-/interrupting timer boundary events:
Added tests to check for correct validation of Timedate.
Added tests to check for correct creation of incident when dynamic inputs are used that have invalid values.
Added tests for the creation and firing of timers. 

## Related issues

https://github.com/camunda/camunda-modeler/issues/3516

closes #5141

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [x] I've reviewed my own code
* [x] I've written a clear changelist description
* [x] I've narrowly scoped my changes
* [x] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.
_Comment: Arguably the linting is incorrect without this, as it does allow all of this even though it isn't supported._

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
_I think my other merge request will adress this, but the modelers need to allow TimeDate as an option for both events._ 
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

Wolgo commented:
I'm all for extending this to sub-processes or other tests if need, but I think it's good to have some feedback before I'm going off in the wrong direction.

Wolgo commented:
I blame the main branch for breaking that smoke test

remcowesterhoud commented:
Thanks for the contribution @Wolgo! I've ran the CI to see if everything still works. I'll do a proper review next week

remcowesterhoud commented on engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/TimerValidationTest.java:
```diff
@@ -177,6 +188,16 @@ private static Stream<Arguments> timerEventsWithExpression() {
                             .intermediateCatchEvent("intermediate-catch-event")
                             .timerWithDurationExpression(expression)
                             .done())),
+            Arguments.of(
+                "intermediate catch event",
+                "date",
+                processBuilder(
+                    expression ->
+                        Bpmn.createExecutableProcess("process")
+                            .startEvent()
+                            .intermediateCatchEvent("intermediate-catch-event")
+                            .timerWithDateExpression(expression)
+                            .done())),
```
❌  We should add a case for the event-sub process here as well.

Wolgo replied:
Thank you very much. I will try to do so. Hope you are feeling well :)

Wolgo replied:
@remcowesterhoud I have added subprocess validation tests and some timer tests. I have limited it to test regarding sub-processes started by a timer, assuming that generic timer or generic sub-processes concerns are tested in those aspects.

remcowesterhoud reviewed (changes requested):
Thanks for the changes @Wolgo! The changes look good so far 🚀  As you mentioned:
> (I guess I should also have done this for subprocesses, oh well)

It would be great if you could extend the PR with this. That way we can actually close the issue instead of having a half completed issue remain 😄

remcowesterhoud commented on engine/src/test/java/io/camunda/zeebe/engine/processing/timer/TimerSubprocessEventTest.java:
```diff
@@ -0,0 +1,187 @@
+import org.junit.Rule;
+import org.junit.Test;
+
+public class TimerSubprocessEventTest {
```
❌ Nice job on the tests 👍 I think we can do this in a simpler way though. We can add a parameter to the [InterruptingEventSubprocessTest)[https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/InterruptingEventSubprocessTest.java] which uses a date time to trigger the timer. This class already has the testcases to verify the cases you have here.

It would be great to also put it in the [NonInterruptingEventSubprocessTest](https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/NonInterruptingEventSubprocessTest.java), but there is two tests that verify the timer is triggered twice, which it won't do when it's supposed to trigger on a specific date time. Maybe you can add an option to the parameters to skip these tests for this case?

Wolgo replied:
Makes a lot of sense, I was looking for a place to put them, but it's so much I must have missed it.

remcowesterhoud reviewed (changes requested):
Great work @Wolgo 👍 I have a suggestion to reuse the tests we already have, so we don't need the new test class here.

There is a two other issues which fail the CI currently:
1. The formatting is off. Please run `mvn spotless:apply` to fix this.
2. Your commits don't adhere to our [Commit Guidelines](https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#commit-message-guidelines). Please have a look at those and rename your commits. An example of a good commit message here would be something along the lines of:

```
test(engine): verify datetime timers in event-subprocesses
```

Let me know when you've processed this feedback and I can go ahead and merge it 🚀

Wolgo commented:
@remcowesterhoud I ran spotless, didn't seem to do anything, but that might have been because i removed the class. 
I fixed the commit messages! (Messing up git is fun at home). 

Hope it will be all right like this!

remcowesterhoud reviewed (approved):
@Wolgo Thanks for the contribution 🚀 All looks good to me!

bors merge

Wolgo commented:
@remcowesterhoud I see it's still failing on formatting, I can't find any way to make mvn spotless:apply work sadly, so If you could maybe run that yourself instead. 😞 I wish I could just fix it

remcowesterhoud commented:
> @remcowesterhoud I see it's still failing on formatting, I can't find any way to make mvn spotless:apply work sadly, so If you could maybe run that yourself instead. 😞 I wish I could just fix it

Hmm interesting. Did you run skip the tests by any chance? Anyways, I've ran it for you, but I had to create a new PR for it #12356

Wolgo commented:
@remcowesterhoud 
![image](https://user-images.githubusercontent.com/15855950/231228191-9343e805-462b-4b88-93d4-24ecb64b79ae.png)
![image](https://user-images.githubusercontent.com/15855950/231228231-29c445d6-27a4-4e86-ba90-5413227d5585.png)

Either one didn't work. But thank you for doing it <3
